tools/rtpsrelay/RelayHandler.cpp

```diff
@@ -288,6 +315,104 @@ bool SpdpHandler::do_normal_processing(const ACE_INET_Addr& a_remote,
   return true;
 }
 
+bool VerticalHandler::parse_message(OpenDDS::RTPS::MessageParser& message_parser,
+                                    ACE_Message_Block* msg,
+                                    bool& is_beacon,
+                                    bool check_submessages)
+{
+  if (!message_parser.parseHeader()) {
+    ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: RelayHandler::parse_message failed to deserialize RTPS header\n"));
+    return false;
+  }
+
+  const auto& header = message_parser.header();
+  OpenDDS::DCPS::RepoId src_guid;
+  std::memcpy(src_guid.guidPrefix, header.guidPrefix, sizeof(OpenDDS::DCPS::GuidPrefix_t));
+  src_guid.entityId = OpenDDS::DCPS::ENTITYID_PARTICIPANT;
+
+  do {
```

Comment:
It seems to be safe to do it this way, but I'm not sure there's any value in having `do`/`while` here instead of just plain `while` (with the same condition).  A message without submessges is not legal, but this should be robust to it anyway.

Reply:
From the code for the MessageParser, it looked like the submessage was not initialized until I called parseSubMessage.  Thus, I didn't want to call hasNextSubmessage until I called parseSubMessage.  Am I missing something?

Reply:
Looks like the previous uses of it rely on `remaining`:
https://github.com/objectcomputing/OpenDDS/blob/acb816282aee2477efd0e65418ff87d09ddf003f/dds/DCPS/security/CryptoBuiltInImpl.cpp#L1028